x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/ClusterDeprecationChecks.java

```diff
@@ -91,6 +96,49 @@ static DeprecationIssue checkTemplatesWithTooManyFields(ClusterState state) {
         return null;
     }
 
+    /**
+     * Check templates that use `_field_names` explicitly, which was deprecated in https://github.com/elastic/elasticsearch/pull/42854
+     * and will throw an error on new indices in 8.0
+     */
+    static DeprecationIssue checkTemplatesWithFieldNamesDisabled(ClusterState state) {
+        Set<String> templatesContainingFieldNames = new HashSet<>();
+        state.getMetaData().getTemplates().forEach((templateCursor) -> {
+            String templateName = templateCursor.key;
+            templateCursor.value.getMappings().forEach((mappingCursor) -> {
+                Map<String, Object> map = XContentHelper.convertToMap(mappingCursor.value.compressedReference(), false, XContentType.JSON)
+                        .v2();
+                // map should contain only type name at this level
```

Comment:
I'm currently assuming that whatever we get here from iterating over the mappings in CompressedXContent contains a type field on the top level. Should also always be one since 6.x if I'm not mistaken

Reply:
> I'm currently assuming that whatever we get here from iterating over the mappings in CompressedXContent contains a type field on the top level.

This is in fact the invariant we try to maintain any time we store a mapping. However, until recently there was a bug where some mappings could be stored without a type (in both indices and index templates): https://github.com/elastic/elasticsearch/pull/45120. So to be extra safe, we could perform the same 'unwrapping' as we do in `MappingMetaData#getSourceAsMap`: https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetaData.java#L141-L146.